File: Ruby/solutions (from jayqui)/06-palindrome.rb
# Write a method that takes a string and returns true if it is a
# palindrome. A palindrome is a string that is the same whether written
# backward or forward. Assume that there are no spaces; only lowercase
# letters will be given.
#
# Difficulty: easy.

### lazy solution
def palindrome?(string)
  string == string.reverse
end

### recursive solution
# def palindrome?(string)
#   return true if string.length == 0 || string.length == 1
#   return false if string[0] != string[-1]
#   palindrome?(string[1..-2])
# end

# These are tests to check that your code is working. After writing
# your solution, they should all print true.

puts('palindrome?("abc") == false: ' + (palindrome?('abc') == false).to_s)
puts('palindrome?("abcba") == true: ' + (palindrome?('abcba') == true).to_s)
puts('palindrome?("z") == true: ' + (palindrome?('z') == true).to_s)

File: Python/problems/06-palindrome.py
# Write a method that takes a string and returns True if it is a
# palindrome. A palindrome is a string that is the same whether written
# backward or forward. Assume that there are no spaces; only lowercase
# letters will be given.
#
# Difficulty: easy.

# ITERATIVE SOLUTION
# import math
# def is_palindrome(string):
#     middle_index = math.ceil(len(string)) - 1

#     index = 0
#     while index < middle_index:
#         if string[index] != string[-1 * (index + 1)]: return False
#         index += 1
#     return True

# RECURSIVE SOLUTION
def is_palindrome(string):

# These are tests to check that your code is working. After writing
# your solution, they should all print True.

print('palindrome?("abc") == False: ' + str(is_palindrome('abc') == False))
print('is_palindrome("abcba") == True: ' + str(is_palindrome('abcba') == True))
print('is_palindrome("z") == True: ' + str(is_palindrome('z') == True))
